### DISCUSSION TITLE
How to adapt `pandoc.Figure` pre-1.3 code (like in `pandoc-ext/diagram.lua`) for 1.4 ?

### DISCUSSION THREAD

**Discussion by fradav** [quarto-dev/quarto-cli#7750] [Q&A]
### Description

Due to some breaking changes in AST for cross-references in 1.4, previous code for producing `pandoc.Figure` should be adapted. 
The actual https://quarto.org/docs/prerelease/1.4/lua_changes.html#support-for-crossreferenceable-elements-in-filters doc for this is a little scarce on this topic, and I can’t figure how to make it for existing `pandoc-ext/diagram.lua` to work.

**Comment by cscheid** [comment]
What specific question do you have?

**Reply by fradav** [reply]
```lua
    -- Store the data in the media bag:
    pandoc.mediabag.insert(fname, imgtype, imgdata)

    -- Create the image object.
    local image = pandoc.Image(dgr_opt.alt, fname, "", dgr_opt['image-attr'])

    -- Create a figure if the diagram has a caption; otherwise return
    -- just the image.
    return dgr_opt.caption and
      pandoc.Figure(
        pandoc.Plain{image},
        dgr_opt.caption,
        dgr_opt['fig-attr']
      ) or
      pandoc.Plain{image}
```

This the code in     https://github.com/pandoc-ext/diagram/blob/main/_extensions/diagram/diagram.lua which works in 1.3, but doesn’t anymore with 1.4. I understand it should be wrapped in a `FloatRefTarget` to be properly handled in the new crossref pipeline.

**Reply by cscheid** [reply]
This is something we still need to add to our docs. You want to use something like

```lua

return quarto.FloatRefTarget {
    identifier = "fig-...",
    type = "Figure",
    content = pandoc.Blocks({...})
    caption_long = pandoc.Blocks({...})
}
```

This filter needs to be run _before_ quarto does its processing, so it's a "before-quarto" filter in 1.4 parlance, and a "pre" filter in 1.3 parlance.
